torch/csrc/jit/interpreter.cpp

```diff
@@ -697,12 +725,21 @@ struct InterpreterStateImpl {
 
     }
   }
+  void registerOutputs(const Instruction& inst, Stack& stack) {
+    for (int i = inst.outputs.size - 1; i >= 0; --i) {
+      int reg = get(inst.outputs, i);
+      registers[reg] = pop(stack);
+      // std::cout << "pop reg[" << reg << "];\n" << registers[reg] << "\n";
+    }
+  }
   // note: it may seem unnecessary to keep the current_pc inside InterpreterState
   // since InterpreterState::run completes the function. However, in the
   // future we will end up with interpreters that can suspend (e.g. for asynchrony)
   // so we keep this design in place eventhough we removed the 'staging'
   // that it was originally used for.
   size_t current_pc = 0;
+  Instruction* current_inst = nullptr;
+  c10::intrusive_ptr<Future> future = c10::make_intrusive<Future>();
```

Comment:
We should avoid allocating the future along the synchronous pathway. Here is my suggestion. We have 2 entry points:

```
void run(Stack& stack) {
    bool suspended = runImpl(stack);
    if(suspended)) {
       future.wait();
      if(num_outputs == 1)
         stack.push(future.value())
      else
        <unpack the tuple, pushing each value onto the stack>
   }

}
intrusive_ptr<Future> runAsync(Stack& stack) {
    if(!future) {
       future = make_instrusive<Future>();
    }
   runImpl(stack);
   return future;
}
// private, replaces current run
bool runImpl(Stack& stack) {
    <do normal run>
    ...
   catch(Suspend& e) {
      if(!future)
         future = make_future(); // we need to ensure this is allocated regardless
      return true; //we suspended and did not push the results to the stack
   }
     ...
    // under the synchronous pathway future will only be allocated on a suspend, 
    // so guard the mark of completion for the case where we never allocate it.
    if(future) {
        <markCompletedLogic>
    }
}
```